File: parsers/ramiroveiculos.py
# parsers/ramiroveiculos.py
import re
from bs4 import BeautifulSoup
from parsers.base_parser import BaseParser

class RamiroVeiculosParser(BaseParser):
    def search(self, query):
        # Ramiro Veículos busca através de /estoque?veiculo=termo
        url = f"{self.base_url}/estoque"
        params = {"veiculo": query}
        
        try:
            html = self.fetch_url(url, params=params)
        except Exception as e:
            print(f"Erro ao acessar {self.name}: {e}")
            return []

        soup = BeautifulSoup(html, "html.parser")
        results = []

        # Encontra os cards de veículos (geralmente sob a classe 'destaque' ou contendo 'card-destaque')
        cards = soup.find_all("div", class_="card-destaque")
        if not cards:
            # Tenta encontrar pela classe 'destaque'
            cards = soup.find_all("div", class_="destaque")

        for card in cards:
            try:
                # 1. URL (Link direto)
                # O card tem vários links. Vamos pegar o primeiro com href
                a_tags = card.find_all("a", href=True)
                link = None
                for a in a_tags:
                    href = a["href"]
                    # Evita links de whatsapp ou links vazios
                    if "veiculo" in href:
                        link = href
                        if not link.startswith("http"):
                            link = f"{self.base_url}/{link.lstrip('/')}"
                        break
                
                if not link and a_tags:
                    # Fallback para qualquer link do card
                    link = a_tags[0]["href"]
                    if link and not link.startswith("http"):
                        link = f"{self.base_url}/{link.lstrip('/')}"

                if not link:
                    continue

                # 2. Preço
                price_tag = card.find(class_="valor-destacadado")
                price = price_tag.get_text(strip=True) if price_tag else None
                
                # Fallback se não encontrar o preço
                if not price:
                    for elem in card.find_all(text=True):
                        if "R$" in elem:
                            cleaned = elem.strip()
                            if len(cleaned) < 25:
                                price = cleaned
                                break
                
                if not price:
                    price = "Sob consulta"

                # 3. Imagem
                # O site Ramiro Veículos usa background-image inline no div com classe 'destaque-imagem'
                image_url = None
                img_div = card.find("div", class_="destaque-imagem")
                if img_div and img_div.get("style"):
                    style = img_div["style"]
                    # Expressão regular para extrair a URL de background: url('...')
                    match = re.search(r"url\(['\"]?(.*?)['\"]?\)", style)
                    if match:
                        image_url = match.group(1)
                        if image_url and not image_url.startswith("http"):
                            image_url = f"{self.base_url}/{image_url.lstrip('/')}"
                
                # Fallback se não for imagem de background
                if not image_url:
                    img_tag = card.find("img")
                    if img_tag:
                        image_url = img_tag.get("src")
                        if image_url and not image_url.startswith("http"):
                            image_url = f"{self.base_url}/{image_url.lstrip('/')}"

                # 4. Título
                # Geralmente no div class="titulo-sub-destaque" com dois <p>
                title = ""
                title_div = card.find(class_="titulo-sub-destaque")
                if title_div:
                    paragraphs = title_div.find_all("p")
                    title_parts = [p.get_text(strip=True) for p in paragraphs if p.get_text(strip=True)]
                    title = " ".join(title_parts)
                
                if not title:
                    # Fallback para o nome do veículo na URL do link
                    # Ex: /veiculo/renault-duster-preto-2022/1409001
                    match = re.search(r"/veiculo/([^/]+)", link)
                    if match:
                        title = match.group(1).replace("-", " ").title()
                    else:
                        title = "Veículo"

                results.append({
                    "title": title,
                    "price": price,
                    "image_url": image_url,
                    "url": link,
                    "dealer_name": self.name
                })
            except Exception as e:
                print(f"Erro ao processar anúncio em {self.name}: {e}")
                continue

        return results

File: api/parsers/ramiroveiculos.js
// api/parsers/ramiroveiculos.js
//
// Parser do site Ramiro Veículos (https://www.ramiroveiculos.com)
// Busca: GET /estoque?veiculo=<query>
// Seletores: .card-destaque ou .destaque
// Imagem: background-image inline em .destaque-imagem

const cheerio = require("cheerio");

const NAME     = "Ramiro Veículos";
const BASE_URL = "https://www.ramiroveiculos.com";

async function search(query, fetchHtml) {
  const url = `${BASE_URL}/estoque?veiculo=${encodeURIComponent(query)}`;
  const html = await fetchHtml(url);
  const $    = cheerio.load(html);
  const results = [];

  const cards = $(".card-destaque").length ? $(".card-destaque") : $(".destaque");

  cards.each((_, card) => {
    try {
      const el = $(card);

      // Link (prioriza href com "veiculo" no caminho)
      let link = null;
      el.find("a[href]").each((_, a) => {
        const href = $(a).attr("href");
        if (href && href.includes("veiculo")) { link = href; return false; }
      });
      if (!link) link = el.find("a[href]").first().attr("href") || null;
      if (link && !link.startsWith("http")) link = `${BASE_URL}/${link.replace(/^\//, "")}`;
      if (!link) return;

      // Imagem — background-image inline em .destaque-imagem
      let image_url = null;
      const style = el.find(".destaque-imagem").attr("style") || "";
      const bgMatch = style.match(/url\(['"]?(.*?)['"]?\)/);
      if (bgMatch) {
        image_url = bgMatch[1];
        if (!image_url.startsWith("http")) image_url = `${BASE_URL}/${image_url.replace(/^\//, "")}`;
      }
      if (!image_url) {
        image_url = el.find("img").first().attr("src") || null;
        if (image_url && !image_url.startsWith("http")) image_url = `${BASE_URL}/${image_url.replace(/^\//, "")}`;
      }

      // Preço
      let price = el.find(".valor-destacadado").text().trim() || null;
      if (!price) {
        el.find("*").each((_, node) => {
          const txt = $(node).text().trim();
          if (txt.includes("R$") && txt.length < 25) { price = txt; return false; }
        });
      }
      if (!price) price = "Sob consulta";

      // Título
      let title = "";
      const titleDiv = el.find(".titulo-sub-destaque");
      if (titleDiv.length) {
        title = titleDiv.find("p").map((_, p) => $(p).text().trim()).get().filter(Boolean).join(" ");
      }
      if (!title) {
        const m = (link || "").match(/\/veiculo\/([^/]+)/);
        title = m ? m[1].replace(/-/g, " ").replace(/\b\w/g, c => c.toUpperCase()) : "Veículo";
      }

      results.push({ title, price, image_url, url: link, dealer_name: NAME });
    } catch (_) {}
  });

  return results;
}

module.exports = { search, name: NAME };
